src/v0/destinations/salesforce/utils.js

```diff
@@ -242,10 +246,226 @@ const getAuthHeader = (authInfo) => {
+    );
+  }
+
+  // If exactly one record is found, check if the lead has been deleted
+  const record = queryResponse.records[0];
+  if (record.IsDeleted === true) {
+    if (record.IsConverted) {
+      throw new NetworkInstrumentationError('The contact has been deleted');
+    }
+    throw new NetworkInstrumentationError('The lead has been deleted.');
+  }
+  if (record.IsConverted && destination.Config.useContactId) {
+    if (record.ConvertedContactId === null) {
+      throw new NetworkInstrumentationError(
+        'The lead is converted but the converted contact id not found',
+      );
+    }
+    return {
+      salesforceType: 'Contact',
+      salesforceId: record.ConvertedContactId,
+    };
+  }
+  return {
+    salesforceType: 'Lead',
+    salesforceId: record.Id,
+  };
+}
+
+async function getSalesforceIdForLeadUsingHttp(email, destination, authInfo, metadata) {
+  const { authorizationData, authorizationFlow } = authInfo;
+  const leadQueryUrl = `${authorizationData.instanceUrl}/services/data/v${SF_API_VERSION}/parameterizedSearch/?q=${email}&sobject=Lead&Lead.fields=id,IsConverted,ConvertedContactId,IsDeleted`;
+
+  // request configuration will be conditional
+  const { processedResponse: processedLeadQueryResponse } = await handleHttpRequest(
+    'get',
+    leadQueryUrl,
+    {
+      headers: getAuthHeader({ authorizationFlow, authorizationData }),
+    },
+    {
+      metadata,
+      destType: 'salesforce',
+      feature: 'transformation',
+      endpointPath: '/parameterizedSearch',
+      requestMethod: 'GET',
+      module: 'router',
+    },
+  );
+
+  if (!isHttpStatusSuccess(processedLeadQueryResponse.status)) {
+    salesforceResponseHandler(
+      processedLeadQueryResponse,
+      `:- during Lead Query`,
+      destination.ID,
+      authorizationFlow,
+    );
+  }
+
+  if (processedLeadQueryResponse.response.searchRecords.length > 0) {
+    // if count is greater than zero, it means that lead exists, then only update it
+    // else the original endpoint, which is the one for creation - can be used
+    const record = processedLeadQueryResponse.response.searchRecords[0];
+    if (record.IsDeleted === true) {
+      if (record.IsConverted) {
+        throw new NetworkInstrumentationError('The contact has been deleted.');
+      } else {
+        throw new NetworkInstrumentationError('The lead has been deleted.');
+      }
+    }
+    if (record.IsConverted && destination.Config.useContactId) {
+      return {
+        salesforceType: 'Contact',
+        salesforceId: record.ConvertedContactId,
+      };
+    }
+    return {
+      salesforceType: 'Lead',
+      salesforceId: record.Id,
+    };
+  }
+  return {
+    salesforceType: 'Lead',
+    salesforceId: undefined,
+  };
+}
+
+async function getSalesforceIdForLead({ email, destination, metadata, stateInfo }) {
+  const whiteListedWorkspaces = process.env.WHITE_LISTED_WORKSPACES_SALESFORCE?.split(',')?.map?.(
+    (s) => s?.trim?.(),
+  );
```

Comment:
Let's refactor this into a utility function.

Reply:
Done